### PR TITLE
ceph: ability to abort orchestration

### LINKED ISSUE #6587: Abort a CephCluster reconcile if the CR is updated while a reconcile is in progress
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
The operator should abort a CephCluster reconcile if the cluster CR is updated. This is particularly an issue during upgrades where both the rook operator and the ceph image are updated at the same time. Both of them will start a new reconcile, and you might need to wait a long time for the second reconcile unnecessarily.

In a test log recently, we saw a 37 minute delay before the second reconcile:
```
# The first reconcile begins and the ceph version is detected:
2020-11-09T13:12:57.509119960Z 2020-11-09 13:12:57.509083 I | ceph-cluster-controller: cluster "openshift-storage": version "14.2.8-91 nautilus" detected for image "quay.io/rhceph-dev/rhceph@sha256:eafd1acb0ada5d7cf93699056118aca19ed7a22e4938411d307ef94048746cc8"

# The cluster CR is updated, which will start the second reconcile as soon as the first reconcile is done. Notice how this happens immediately after the first reconcile started, so we have to wait a long time.
2020-11-09T13:12:57.966652084Z 2020-11-09 13:12:57.966596 I | ceph-spec: CR has changed for "ocs-storagecluster-cephcluster". diff=  v1.ClusterSpec{

# The second reconcile finally starts 37 minutes later
2020-11-09 13:49:57.272222 I | ceph-cluster-controller: detected ceph image version: "14.2.8-111 nautilus"
```

We are printing that the cluster spec changed [here](https://github.com/rook/rook/blob/23108cc94afdebc8f4ab144130a270b1e4ffd94e/pkg/operator/ceph/controller/predicate.go#L275), but then we wait (potentially a long time) before we ever act on it.

The operator should check at various intervals to see if the cluster has been updated and therefore should abort the reconcile in progress. For example, we should check when:
- After the mons are updated
- In between each osd being updated, such as when we are checking if an osd is ok-to-stop. This tends to take the longest during a reconcile, and the reconcile can be aborted with low risk during OSD reconcile.

**What is use case behind this feature:**
More responsive reconcile, particularly during upgrades

### PR DESCRIPTION
**Pull request by leseb**
**Description of your changes:**

We can now prioritize orchestrations on certain event. Today only two
events will cancel on-going orchestrations (if any):

* request for cluster deletion
* request for cluster upgrade

If one of the two are caught by the watcher we will cancel the on-going
orchestration.
For that we implemented a simple approach based on check points, where
we will check for a cancellation request in certain part of the
orchestration. Mainly before each mons/mgr/osds orchestration loops.

This solution is not perfect, but we are waiting for the
controller-runtime to release its 0.7 version which will embed context
support. With that we will be able to cancel reconciles more precisely
and rapidly.

Operator log example:

```
2020-11-25 12:59:12.986264 I | ceph-cluster-controller: done reconciling ceph cluster in namespace "rook-ceph"











2020-11-25 13:07:33.776947 I | ceph-cluster-controller: CR has changed for "rook-ceph". diff=  v1.ClusterSpec{
        CephVersion: v1.CephVersionSpec{
                Image:            "ceph/ceph:v15.2.5",
-               AllowUnsupported: true,
+               AllowUnsupported: false,
        },
        DriveGroups: nil,
        Storage:     {UseAllNodes: true, Selection: {UseAllDevices: &true}},
        ... // 20 identical fields
  }
2020-11-25 13:07:33.777039 I | ceph-cluster-controller: reconciling ceph cluster in namespace "rook-ceph"
2020-11-25 13:07:33.785088 I | op-mon: parsing mon endpoints: a=10.107.242.49:6789,b=10.109.71.30:6789,c=10.98.93.224:6789
2020-11-25 13:07:33.788626 I | ceph-cluster-controller: detecting the ceph image version for image ceph/ceph:v15.2.5...
2020-11-25 13:07:35.280789 I | ceph-cluster-controller: detected ceph image version: "15.2.5-0 octopus"
2020-11-25 13:07:35.280806 I | ceph-cluster-controller: validating ceph version from provided image
2020-11-25 13:07:35.285888 I | op-mon: parsing mon endpoints: a=10.107.242.49:6789,b=10.109.71.30:6789,c=10.98.93.224:6789
2020-11-25 13:07:35.287828 I | cephclient: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-11-25 13:07:35.288082 I | cephclient: generated admin config in /var/lib/rook/rook-ceph
2020-11-25 13:07:35.621625 I | ceph-cluster-controller: cluster "rook-ceph": version "15.2.5-0 octopus" detected for image "ceph/ceph:v15.2.5"
2020-11-25 13:07:35.642688 I | op-mon: start running mons
2020-11-25 13:07:35.646323 I | op-mon: parsing mon endpoints: a=10.107.242.49:6789,b=10.109.71.30:6789,c=10.98.93.224:6789
2020-11-25 13:07:35.654070 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.107.242.49:6789","10.109.71.30:6789","10.98.93.224:6789"]}] data:a=10.107.242.49:6789,b=10.109.71.30:6789,c=10.98.93.224:6789 mapping:{"node":{"a":{"Name":"minikube","Hostname":"minikube","Address":"192.168.39.3"},"b":{"Name":"minikube","Hostname":"minikube","Address":"192.168.39.3"},"c":{"Name":"minikube","Hostname":"minikube","Address":"192.168.39.3"}}} maxMonId:2]
2020-11-25 13:07:35.868253 I | cephclient: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-11-25 13:07:35.868573 I | cephclient: generated admin config in /var/lib/rook/rook-ceph
2020-11-25 13:07:37.074353 I | op-mon: targeting the mon count 3
2020-11-25 13:07:38.153435 I | op-mon: checking for basic quorum with existing mons
2020-11-25 13:07:38.178029 I | op-mon: mon "a" endpoint is [v2:10.107.242.49:3300,v1:10.107.242.49:6789]
2020-11-25 13:07:38.670191 I | op-mon: mon "b" endpoint is [v2:10.109.71.30:3300,v1:10.109.71.30:6789]
2020-11-25 13:07:39.477820 I | op-mon: mon "c" endpoint is [v2:10.98.93.224:3300,v1:10.98.93.224:6789]
2020-11-25 13:07:39.874094 I | op-mon: saved mon endpoints to config map map[csi-cluster-config-json:[{"clusterID":"rook-ceph","monitors":["10.107.242.49:6789","10.109.71.30:6789","10.98.93.224:6789"]}] data:a=10.107.242.49:6789,b=10.109.71.30:6789,c=10.98.93.224:6789 mapping:{"node":{"a":{"Name":"minikube","Hostname":"minikube","Address":"192.168.39.3"},"b":{"Name":"minikube","Hostname":"minikube","Address":"192.168.39.3"},"c":{"Name":"minikube","Hostname":"minikube","Address":"192.168.39.3"}}} maxMonId:2]
2020-11-25 13:07:40.467999 I | cephclient: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-11-25 13:07:40.469733 I | cephclient: generated admin config in /var/lib/rook/rook-ceph
2020-11-25 13:07:41.071710 I | cephclient: writing config file /var/lib/rook/rook-ceph/rook-ceph.config
2020-11-25 13:07:41.078903 I | cephclient: generated admin config in /var/lib/rook/rook-ceph
2020-11-25 13:07:41.125233 I | op-mon: deployment for mon rook-ceph-mon-a already exists. updating if needed
2020-11-25 13:07:41.327778 I | op-k8sutil: updating deployment "rook-ceph-mon-a" after verifying it is safe to stop
2020-11-25 13:07:41.327895 I | op-mon: checking if we can stop the deployment rook-ceph-mon-a
2020-11-25 13:07:44.045644 I | op-k8sutil: finished waiting for updated deployment "rook-ceph-mon-a"
2020-11-25 13:07:44.045706 I | op-mon: checking if we can continue the deployment rook-ceph-mon-a
2020-11-25 13:07:44.045740 I | op-mon: waiting for mon quorum with [a b c]
2020-11-25 13:07:44.109159 I | op-mon: mons running: [a b c]
2020-11-25 13:07:44.474596 I | op-mon: Monitors in quorum: [a b c]
2020-11-25 13:07:44.478565 I | op-mon: deployment for mon rook-ceph-mon-b already exists. updating if needed
2020-11-25 13:07:44.493374 I | op-k8sutil: updating deployment "rook-ceph-mon-b" after verifying it is safe to stop
2020-11-25 13:07:44.493403 I | op-mon: checking if we can stop the deployment rook-ceph-mon-b
2020-11-25 13:07:47.135524 I | op-k8sutil: finished waiting for updated deployment "rook-ceph-mon-b"
2020-11-25 13:07:47.135542 I | op-mon: checking if we can continue the deployment rook-ceph-mon-b
2020-11-25 13:07:47.135551 I | op-mon: waiting for mon quorum with [a b c]
2020-11-25 13:07:47.148820 I | op-mon: mons running: [a b c]
2020-11-25 13:07:47.445946 I | op-mon: Monitors in quorum: [a b c]
2020-11-25 13:07:47.448991 I | op-mon: deployment for mon rook-ceph-mon-c already exists. updating if needed
2020-11-25 13:07:47.462041 I | op-k8sutil: updating deployment "rook-ceph-mon-c" after verifying it is safe to stop
2020-11-25 13:07:47.462060 I | op-mon: checking if we can stop the deployment rook-ceph-mon-c
2020-11-25 13:07:48.853118 I | ceph-cluster-controller: CR has changed for "rook-ceph". diff=  v1.ClusterSpec{
        CephVersion: v1.CephVersionSpec{
-               Image:            "ceph/ceph:v15.2.5",
+               Image:            "ceph/ceph:v15.2.6",
                AllowUnsupported: false,
        },
        DriveGroups: nil,
        Storage:     {UseAllNodes: true, Selection: {UseAllDevices: &true}},
        ... // 20 identical fields
  }
2020-11-25 13:07:48.853140 I | ceph-cluster-controller: upgrade requested, cancelling any ongoing orchestration
2020-11-25 13:07:50.119584 I | op-k8sutil: finished waiting for updated deployment "rook-ceph-mon-c"
2020-11-25 13:07:50.119606 I | op-mon: checking if we can continue the deployment rook-ceph-mon-c
2020-11-25 13:07:50.119619 I | op-mon: waiting for mon quorum with [a b c]
2020-11-25 13:07:50.130860 I | op-mon: mons running: [a b c]
2020-11-25 13:07:50.431341 I | op-mon: Monitors in quorum: [a b c]
2020-11-25 13:07:50.431361 I | op-mon: mons created: 3
2020-11-25 13:07:50.734156 I | op-mon: waiting for mon quorum with [a b c]
2020-11-25 13:07:50.745763 I | op-mon: mons running: [a b c]
2020-11-25 13:07:51.045108 I | op-mon: Monitors in quorum: [a b c]
2020-11-25 13:07:51.054497 E | ceph-cluster-controller: failed to reconcile. failed to reconcile cluster "rook-ceph": failed to configure local ceph cluster: failed to create cluster: CANCELLING CURRENT ORCHESTATION
2020-11-25 13:07:52.055208 I | ceph-cluster-controller: reconciling ceph cluster in namespace "rook-ceph"
2020-11-25 13:07:52.070690 I | op-mon: parsing mon endpoints: a=10.107.242.49:6789,b=10.109.71.30:6789,c=10.98.93.224:6789
2020-11-25 13:07:52.088979 I | ceph-cluster-controller: detecting the ceph image version for image ceph/ceph:v15.2.6...
2020-11-25 13:07:53.904811 I | ceph-cluster-controller: detected ceph image version: "15.2.6-0 octopus"
2020-11-25 13:07:53.904862 I | ceph-cluster-controller: validating ceph version from provided image
```

Closes: #6587

**Which issue is resolved by this Pull Request:**
Resolves #6587

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

**Comment by leseb** [comment]
More testing to do on my end, but good for an initial review.

**Review by BlaineEXE — COMMENTED** [review]
In the commit message. `Also removing one used variable.` Did you mean "unused" variable?

Edit: I was hoping this would be attached to the commit I was looking at but I guess not. This is the first commit message.

**Review by BlaineEXE — COMMENTED** [review]
With adding a dedicated predicate for the `CephCluster` object, the thing I have concerns about is what happens if any of the other cases (CephObjectStore, CephBlockPool, CephFilesystem, CephNFS, CephRBDMirror, etc.) have CRDs created before the CephCluster is created. Assuming I don't find anything wrong, I will likely not request changes and trust that you'll test this before merging.

File: pkg/clusterd/context.go [hunk 1]
```diff
@@ -66,4 +67,7 @@ type Context struct {
 
 	// The local devices detected on the node
 	Devices []*sys.LocalDisk
+
+	// OrchestrationCanceller manages the orchestration and its possible cancellation
+	OrchestrationCanceller *abool.AtomicBool
```
Comment (BlaineEXE) [inline pkg/clusterd/context.go:72]: As a nit, the methods on this are `Set`, `Unset`, and `IsSet`, and I don't read "(un)set canceller" as having the right semantic meaning. I think renaming this to "RequestCancelOrchestration" or "OrchestrationCancellationRequest" or something similar makes the method use more grammatically specific to what's going on in the code.

File: pkg/operator/ceph/cluster/predicate.go [hunk 2]
```diff
@@ -133,6 +134,13 @@ func watchControllerPredicate(rookContext *clusterd.Context) predicate.Funcs {
 				diff := cmp.Diff(objOld.Spec, objNew.Spec, resourceQtyComparer)
 				if diff != "" {
 					logger.Infof("CR has changed for %q. diff=%s", objNew.Name, diff)
+
+					// If spec Images are different we cancel any on-going orchestration
+					if objNew.Spec.CephVersion.Image != objOld.Spec.CephVersion.Image {
+						logger.Info("upgrade requested, cancelling any on-going orchestration")
```
Comment (BlaineEXE) [inline pkg/operator/ceph/cluster/predicate.go:140]: nit: "on-going" -> "ongoing"

**Review by BlaineEXE — APPROVED** [review]
Two small nits.

File: pkg/clusterd/context.go (hunk 1 above)
Comment (leseb) [reply pkg/clusterd/context.go:72]: Will go with `RequestCancelOrchestration` thanks for the suggestion.

**Comment by leseb** [comment]
> With adding a dedicated predicate for the `CephCluster` object, the thing I have concerns about is what happens if any of the other cases (CephObjectStore, CephBlockPool, CephFilesystem, CephNFS, CephRBDMirror, etc.) have CRDs created before the CephCluster is created. Assuming I don't find anything wrong, I will likely not request changes and trust that you'll test this before merging.

Conceptually, it makes more sense to have a dedicated predicate for each controller. I had previously homogenized the code to end up with a single one but because this cancellation approach is only for the CephCluster I didn't want to update all the other controller's definitions for something we might never use.
Extracting the predicate does not change the behavior of the other controllers, they will behave the same as they were. If a child is injected before a CephCluster one, the child controller will respond normally and wait for the CephCluster to appear so it can reconcile.

Does that clarify?
Thanks!

**Comment by BlaineEXE** [comment]
> > With adding a dedicated predicate for the `CephCluster` object, the thing I have concerns about is what happens if any of the other cases (CephObjectStore, CephBlockPool, CephFilesystem, CephNFS, CephRBDMirror, etc.) have CRDs created before the CephCluster is created. Assuming I don't find anything wrong, I will likely not request changes and trust that you'll test this before merging.
> 
> Conceptually, it makes more sense to have a dedicated predicate for each controller. I had previously homogenized the code to end up with a single one but because this cancellation approach is only for the CephCluster I didn't want to update all the other controller's definitions for something we might never use.
> Extracting the predicate does not change the behavior of the other controllers, they will behave the same as they were. If a child is injected before a CephCluster one, the child controller will respond normally and wait for the CephCluster to appear so it can reconcile.
> 
> Does that clarify?
> Thanks!

Yes! That seems obvious now that you have said it. I still haven't had coffee yet today. 😬 ☕

**Pull request merged**